homeassistant/components/octoprint.py

```diff
@@ -11,43 +11,121 @@
 import voluptuous as vol
 from aiohttp.hdrs import CONTENT_TYPE
 
-from homeassistant.const import CONF_API_KEY, CONF_HOST, CONTENT_TYPE_JSON
+from homeassistant.components.discovery import SERVICE_OCTOPRINT
+from homeassistant.const import (
+    CONF_API_KEY, CONF_HOST, CONTENT_TYPE_JSON, CONF_NAME, CONF_PORT,
+    CONF_SSL, TEMP_CELSIUS, CONF_MONITORED_CONDITIONS, CONF_SENSORS,
+    CONF_BINARY_SENSORS)
+from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.util import slugify as util_slugify
 
 _LOGGER = logging.getLogger(__name__)
 
 DOMAIN = 'octoprint'
 CONF_NUMBER_OF_TOOLS = 'number_of_tools'
 CONF_BED = 'bed'
+DEFAULT_NAME = 'OctoPrint'
+
+
+def has_all_unique_names(value):
+    """Validate that printers have an unique name."""
+    names = []
+    for printer in value:
+        name = printer['name'] if 'name' in printer else DEFAULT_NAME
```

Comment:
```py
name = printer.get('name', DEFAULT_NAME)
```

Reply:
Even better, move the call to `has_all_unique_names` in the config schema after the main part in the schema, ie after `[vol.Schema({...}]`. Then default values will have been populated in the dict.

Reply:
I just tried that but unfortunately it didn't gave me a very usable result: If I move the call to `has_all_uniques_names` to the next `vol.Schema`.

Going from:
```python
  DOMAIN: vol.All(cv.ensure_list, has_all_unique_names, [vol.Schema({
```
to (if I understood correctly):
```python
  DOMAIN: vol.All(cv.ensure_list, [vol.Schema(has_all_unique_names, {
```

`has_all_uniques_names` gets called once for each entry, instead of being called only once with all entries. It makes verifying all the names at once for uniqueness more difficult.

Reply:
No it should be:
```py
DOMAIN: vol.All(cv.ensure_list, [vol.Schema({...})], has_all_unique_names),
```
I might have forgotten a parenthesis in my earlier comment.

Reply:
Got it, that makes more sense. I changed it that way and it indeed simplified the custom validator.